### DISCUSSION TITLE
Does tools listed under 'Cached Tools' in Ubuntu2004-Readme.md mean it is installed?

### DISCUSSION THREAD

**Discussion by jay0129** [actions/virtual-environments#8193] [General questions]
I use latest image but I can only find Python 3.8.10.

Does it mean that these versions are eligible to be used by actions/setup-python?

Reference:
https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md#cached-tools

**Comment by vpolikarpov-akvelon** [comment]
Hey @jay0129. These tools are installed indeed, but to the special directory that is referenced by environment variable `$AGENT_TOOLSDIRECTORY`.

When you use `actions/setup-python` it checks whether version you require is in that directory or not and either quickly bootstraps environment using cached version or downloads Python from the internet.

If you want to list all versions of Python that are preinstalled on runner you may run `ls -l $AGENT_TOOLSDIRECTORY/Python`. If you want to learn how to activate one of these cached Python versions you may examine source code of `setup-python` action: https://github.com/actions/setup-python/tree/main/src
